### DISCUSSION TITLE
[Tooling] : Code completion with macros doesn't seem to work with current rust-analyzer of vscode

### DISCUSSION THREAD

**Discussion by bebenlebricolo** [juhaku/utoipa#654] [Q&A] [answered]
Hello guys, 

I'm trying to wrap my head around the utoipa crate and trying to tweak examples (the [rocket-0_4-hello](https://github.com/juhaku/utoipa/tree/master/examples/rocket-0_4-hello) to be precise) and whenever I try to change the smallest thing within the ` #[openapi(paths(hello))]`  or `#[utoipa::path(...)]`, rust-analyzer cannot cope with it and I'm left with nothing.

#### Trying to get the code completion working (...)
![image](https://github.com/juhaku/utoipa/assets/29211687/b9e85383-85dd-4772-bbbd-b8d3bd7c3fc5)

This is a tooling issue, very much probably on my side but I don't understand why this thing is not doing the proper stuff .
So first question : ***what's your setup and is code completion working inside macros instructions ?***

As I understand it right now, the ` [#derive(OpenApi)]` macro on top of the ` struct ApiDoc` expands and lets rustc know about the other available macros, which only become available if the `derive` one is used (thanks for the compiler warning by the way, it's what helped me understand what was going on).

For a little bit more context (if it even has some importance) :
I'm a C++ / C# dev and I'm quite familiar with the AspNet / Swagger frameworks ; so I understand most of the principles under the hood, but I lack some proper Rust seasoning and fighting my tools like that does not help :smile: .

It forces me to get through the documentation ( noticeably this part : [path-attributes](https://docs.rs/utoipa/latest/utoipa/attr.path.html#path-attributes)) and reading the utoipa source code to understand what's going on.

#### Same thing here on the very top of the file :
![image](https://github.com/juhaku/utoipa/assets/29211687/4e1b452f-d50b-4a03-8647-a41d406b295a)
As you can see, I can't get the tools to give me the slightest bits of context about this, the rust-analyzer itself is completely fooled ! No context, can't navigate to the lib/doc, etc.. :confused: 

Bottom line being : is it an issue with the rust-analyzer or I'm missing something (probably ...)

Thanks for reading this guys :smile:

**Comment by juhaku** [comment, marked as answer]
> So first question : __what's your setup and is code completion working inside macros instructions ?__

Make sure you are using `rust_analyzer` and for debuggin you might want to use codellb (not needed unless you plan to run apps through the debugger). Both of those can be found from extensions for VSCode.

Then you might want to check that `rust_analyzer` has sensible settings, ones that I am using: _(This config is in lua, but the settings are same for VSCode as well)_ https://rust-analyzer.github.io/manual.html#configuration
```lua
imports = {
    granularity = {
        group = "module",
    },
    prefix = "self",
},
cargo = {
    buildScripts = {
        enable = true,
    },
    features = {},
    noDefaultFeatures = false,
},
procMacro = {
   enable = true,
   attributes = {
        enable = true,
   },
},
checkOnSave = {
    command = "clippy",
}
```
Some of the above configs are default options, but procMacro related settings are not enabled by default and save command is not set to clippy by default.

Then to the actual point, as what comes to the proc macros or macros in general, rust via `rust_analyzer` is in general not able to "autocomplete" the proc macro attributes. This is a still a limitation unfortunately. But it can some times auto complete some values for it's fields, but this purely depends on how the proc macro has been internally implemented. But when setup works it should give you possibility to see and navigate to the lib docs.

Also note that the rocket 0.4 explicitly needs nightly toolchain to be present and it cannot be used / run without it. So you might also want to install it with `rustup toolchain install nightly`. To start building the rocket hello 0.4 example from scratch remove the `Cargo.lock` and run `cargo clean` before building the app and restart your VSCode. This should help you to get started.

Cheers
Juha

**Reply by bebenlebricolo** [reply]
Hi there, thank you very much for the detailed answer.
In the meantime I browsed through the source code and had a grasp about how the proc_macros are generated, and thanks to the documentation of utoipa I managed to get something working as I expected.

So it essentially boils down to being a limitation of `rust_analyzer` for now IMHO (completion-wize, I mean, the rest is due to my still pretty poor grasp of the whole rust ecosystem :smile: ).
This is how far I went with the security tweaks for now : [main.rs](https://github.com/bebenlebricolo/rust-web-service-test/blob/main/src/main.rs).
It's pretty bare and it was solely for me to understand how both rocket and utoipa work together, I used some examples from here and there from the examples provided in this repo + some other suggestions coming from the doc itself.

Thanks again for the reply, and hands down for the development effort you've put into that crate (especially the documentation, once you get into it, there are lots of invaluable examples in it).
